tetragonpod/controller/tetragonpod_controller.go

```diff
@@ -0,0 +1,194 @@
+func deleteTetragonPod(ctx context.Context, r *TetragonPodReconciler, req ctrl.Request, l logr.Logger) (ctrl.Result, error) {
+	l.Info("Deleting the corresponding tetragonPod resource")
+
+	// create the pod to be deleted
+	tetragonPodToDelete := &ciliumiov1alpha1.TetragonPod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+		},
+	}
+
+	// Check if corresponding tetragonPod resource exists
+	if err := r.Get(ctx, req.NamespacedName, tetragonPodToDelete); err != nil {
+		l.Error(err, "Corresponding tetragonPod Resource does not exist")
+		return ctrl.Result{}, err
+	}
+
+	// Delete the Pod
+	if err := r.Delete(ctx, tetragonPodToDelete); err != nil {
+		l.Error(err, "Failed to delete the tetragonPod resource")
+		return ctrl.Result{}, err
+	}
+
+	// Pod deleted successfully.
+	l.Info("TetragonPod deleted successfully")
+	return ctrl.Result{}, nil
+}
+
+// updateTetragonPod updates the corresponding tetragonPod resource.
+func updateTetragonPod(ctx context.Context, r *TetragonPodReconciler, l logr.Logger, pod *corev1.Pod, tetragonPod *ciliumiov1alpha1.TetragonPod) (ctrl.Result, error) {
+	l.Info("Updating the corresponding tetragonPod resource")
+
+	// get the new IP address of the Pod.
+	newIP := ciliumiov1alpha1.PodIP{IP: pod.Status.PodIP}
+	tetragonPod.Status.PodIP = newIP.IP
+	// add the new IP to the beginning of the array.
+	tetragonPod.Status.PodIPs = append([]ciliumiov1alpha1.PodIP{newIP}, tetragonPod.Status.PodIPs...)
+
+	// Update the tetragonPod resource.
+	if err := r.Update(ctx, tetragonPod); err != nil {
+		l.Error(err, "Failed to update the tetragonPod resource")
+		return ctrl.Result{}, err
+	}
+	l.Info("TetragonPod updated successfully")
+	return ctrl.Result{}, nil
+}
+
+// checkIfAlreadyExists checks if the corresponding tetragonPod resource already exists.
+func checkIfAlreadyExists(ctx context.Context, req ctrl.Request, r *TetragonPodReconciler, l logr.Logger) (*ciliumiov1alpha1.TetragonPod, bool) {
+
+	// check if the tetragonPod already exists for that pod.
+	tetragonPod, err := getTetragonPod(ctx, req, r, l)
+	if err != nil {
+		return nil, false
+	}
+	return tetragonPod, true
+}
+
+// createTetragonPod creates a tetragonPod resource.
+func createTetragonPod(ctx context.Context, pod *corev1.Pod, r *TetragonPodReconciler, l logr.Logger) (ctrl.Result, error) {
+	// create the tetragon pod since the pod exists
+	if pod.Status.Phase == corev1.PodRunning {
+		l.Info("Creating the corresponding tetragonPod resource")
+		// get the new IP address of the Pod.
+		newIP := ciliumiov1alpha1.PodIP{IP: pod.Status.PodIP}
+		tetragonPod := &ciliumiov1alpha1.TetragonPod{
+
+			// create the tetragonPod with the same name and namespace as the pod.
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      pod.Name,
+				Namespace: pod.Namespace,
+			},
+			Status: ciliumiov1alpha1.TetragonPodStatus{
+				PodIP:  newIP.IP,
+				PodIPs: make([]ciliumiov1alpha1.PodIP, 0),
+			},
+		}
```

Comment:
extract this logic into a separate function so that you can unit test it.

Reply:
done